### PR TITLE
Allow enums to be serialized by SessionStateTempDataProvider

### LINKED ISSUE #3141: TempData support for Enum types
`SessionStateTempDataProvider` does not support `Enum` types but enums always can be converted to the underlying type and serialized to JSON. May be temp data provider should support enums?

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #3141

### CODE REVIEW

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ViewFeatures/SessionStateTempDataProviderTest.cs [hunk 1]
```diff
@@ -185,7 +186,7 @@ public void SaveAndLoad_IntCanBeStoredAndLoaded()
             var TempData = testProvider.LoadTempData(context);
 
             // Assert
-            var intVal = Convert.ToInt32(TempData["int"]);
+            var intVal = Assert.IsType<int>(TempData["int"]);
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ViewFeatures/SessionStateTempDataProviderTest.cs]: @Eilon, casting for `int`s was currently broken.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/SessionStateTempDataProvider.cs [hunk 2]
```diff
@@ -128,6 +129,16 @@ public class SessionStateTempDataProvider : ITempDataProvider
                             throw new InvalidOperationException(message);
                         }
                     }
+                    else if (item.Value is long)
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/SessionStateTempDataProvider.cs]: Thoughts?

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ViewFeatures/SessionStateTempDataProviderTest.cs (hunk 1 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ViewFeatures/SessionStateTempDataProviderTest.cs]: What breaks if `int`s round-trip as `long`s with the same numeric value?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/SessionStateTempDataProvider.cs (hunk 2 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/SessionStateTempDataProvider.cs]: Meh. Unclear why we'd guess the original type was `int` and not `short` et cetera.

**Comment by dougbu** [comment]
:watch:

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ViewFeatures/SessionStateTempDataProviderTest.cs (hunk 1 above)
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ViewFeatures/SessionStateTempDataProviderTest.cs]: The casting.
```C#
TempData["foo"] = 1;
...
var id = (int)TempData["foo"];
```

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/SessionStateTempDataProvider.cs (hunk 2 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/SessionStateTempDataProvider.cs]: You're far more likely to use `int`s than any other data type.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ViewFeatures/SessionStateTempDataProviderTest.cs (hunk 1 above)
Comment (Eilon) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ViewFeatures/SessionStateTempDataProviderTest.cs]: Right, can't unbox a boxed `long` to an `int`.

**Comment by Eilon** [comment]
Looks :shipit: to me...

**Pull request closed without merging**